systemd/systemd.go

```diff
@@ -1389,8 +1403,73 @@ X-SnapdOrigin={{.}}
 {{- end}}
 `
 
+// We want kernel-modules components to be mounted before modules are
+// loaded (that is before systemd-{udevd,modules-load}).
+const kernelModulesMountUnitTmpl = `[Unit]
+Description={{.Description}}
+DefaultDependencies=no
+After=systemd-remount-fs.service
+Before=systemd-udevd.service systemd-modules-load.service
+Before=umount.target
+Conflicts=umount.target
+
+[Mount]
+What={{.What}}
+Where={{.Where}}
+Type={{.Fstype}}
+Options={{join .Options ","}}
+LazyUnmount=yes
+
+[Install]
+WantedBy=sysinit.target
+{{- with .Origin}}
+X-SnapdOrigin={{.}}
+{{- end}}
+`
+
+// These units form the kernel modules/firmware trees and same as
+// kernel-modules need to be mounted early, but after kernel-modules
+// as they use their content.
+const kernelTreeMountUnitTmpl = `[Unit]
+Description={{.Description}}
+DefaultDependencies=no
+After=systemd-remount-fs.service
+After={{mountedComp .What}}
+Before=systemd-udevd.service systemd-modules-load.service
+Before=umount.target
+Conflicts=umount.target
+
+[Mount]
+What={{.What}}
+Where={{.Where}}
+Type={{.Fstype}}
+Options={{join .Options ","}}
+LazyUnmount=yes
+
+[Install]
+WantedBy=sysinit.target
+{{- with .Origin}}
+X-SnapdOrigin={{.}}
+{{- end}}
+`
+
 var templateFuncs = template.FuncMap{"join": strings.Join}
-var parsedMountUnitTemplate = template.Must(template.New("unit").Funcs(templateFuncs).Parse(mountUnitTemplate))
+var parsedNormalMountUnitTmpl = template.Must(template.New("unit").Funcs(templateFuncs).Parse(normalMountUnitTmpl))
+var parsedKernelModulesMountUnitTmpl = template.Must(template.New("unit").Funcs(templateFuncs).Parse(kernelModulesMountUnitTmpl))
+
+var kernelTreeTemplateFuncs = template.FuncMap{
+	"join": strings.Join,
+	"mountedComp": func(what string) (string, error) {
+		// First 5 elements need to match
+		// /run/mnt/kernel-modules/<kernel_ver>/<component>
+		dirs := strutil.SplitNonEmtpy(what, string(os.PathSeparator))
```

Comment:
If perfix is of concern, then you can also do this:
```
clean := filepath.Clean(what)
if !strings.HasPrefix(clean, "/run/mnt/kernel-modules/") || len(strings.Split(clean)) != 5 {
    // expecting 5 components starting with known prefix
    return "", fmt.Errorf("..")
}
```

Reply:
I've added an additional check, thanks for the suggestion